test/e2e/apiextensions/composition_test.go

```diff
@@ -0,0 +1,735 @@
+									   "name": "nop-example-resource"
+									}
+								}
+							}`)},
+							},
+							{
+								Name: pointer.StringPtr("nopinstance2"),
+								Base: runtime.RawExtension{Raw: []byte(`{
+									"apiVersion": "nop.crossplane.io/v1alpha1",
+									"kind": "NopResource",
+									"spec": {
+										"forProvider": {
+										   "conditionAfter": [
+											  {
+												 "conditionType": "Ready",
+												 "conditionStatus": "False",
+												 "time": "0s"
+											  },
+											  {
+												 "conditionType": "Ready",
+												 "conditionStatus": "True",
+												 "time": "20s"
+											  },
+											  {
+												 "conditionType": "Synced",
+												 "conditionStatus": "False",
+												 "time": "0s"
+											  },
+											  {
+												 "conditionType": "Synced",
+												 "conditionStatus": "True",
+												 "time": "20s"
+											  }
+										   ]
+										},
+										"writeConnectionSecretToRef": {
+										   "namespace": "crossplane-system",
+										   "name": "nop-example-resource"
+										}
+									}
+								}`)},
+							},
+						},
+					},
+				}
+
+				if err := c.Create(ctx, comp); err != nil {
+					t.Fatalf("Create composition %q: %v", comp.GetName(), err)
+				}
+				t.Logf("Created composition %q", comp.GetName())
+
+				t.Cleanup(func() {
+					t.Logf("Cleaning up composition %q.", comp.GetName())
+					if err := c.Get(ctx, types.NamespacedName{Name: comp.GetName()}, comp); resource.IgnoreNotFound(err) != nil {
+						t.Fatalf("Get composition %q: %v", comp.GetName(), err)
+					}
+					if err := c.Delete(ctx, comp); resource.IgnoreNotFound(err) != nil {
+						t.Fatalf("Delete composition %q: %v", comp.GetName(), err)
+					}
+					t.Logf("Deleted composition %q", comp.GetName())
+				})
+
+				nopRes := schema.GroupVersionResource{Group: "nop.example.org", Version: "v1alpha1", Resource: "nopresources"}
+
+				nopresource := &unstructured.Unstructured{
+					Object: map[string]interface{}{
+						"apiVersion": "nop.example.org/v1alpha1",
+						"kind":       "NopResource",
+						"metadata": map[string]interface{}{
+							"name": "nop-example",
+						},
+						"spec": map[string]interface{}{
+							"coolField": "example",
+						},
+					},
+				}
+
+				res, err := dc.Resource(nopRes).Namespace("default").Create(context.TODO(), nopresource, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("Create nop-example %q: %v", nopresource.GetName(), err)
+				}
+
+				t.Logf("Created nop-example %q", res.GetName())
+
+				if err := wait.PollImmediate(10*time.Second, 90*time.Second, func() (done bool, err error) {
+					d, err := dc.Resource(nopRes).Namespace("default").Get(context.TODO(), "nop-example", metav1.GetOptions{})
+					if err != nil {
+						t.Fatalf("Get nop-res %q: %v", nopresource.GetName(), err)
+					}
+
+					conditionList, found, err := unstructured.NestedSlice(d.Object, "status", "conditions")
```

Comment:
One thing you could do here to make these unstructured objects easier to work with is use [the `composed.Unstructured` wrapper struct in crossplane-runtime](https://github.com/crossplane/crossplane-runtime/blob/36fc69eff96ecb5856f156fec077ed3f3c3b30b1/pkg/resource/unstructured/composed/composed.go#L60). Once you construct it, you can utilize the methods it offers, such as [`GetCondition`](https://github.com/crossplane/crossplane-runtime/blob/36fc69eff96ecb5856f156fec077ed3f3c3b30b1/pkg/resource/unstructured/composed/composed.go#L70).

Reply:
Great pointer @hasheddan, thanks for sharing that! that programming model looks nice :)

Reply:
Thank you @hasheddan! That suggestion really made the code look concise and clean. 🎉